### ISSUE TITLE
Measure binary size impact of implicit caller location

### ISSUE THREAD

**Issue by anp** [rust-lang/rust#70579]
The largest unanswered questions about [implicit caller location](https://github.com/rust-lang/rust/issues/47809) are about the binary size impact of the feature and how to offer users control over it.

There are three ways I see that `#[track_caller]` inflates binaries:

1. code bloat from panicking branches that were previously fused
2. encoding the `Location` structs themselves
3. encoding the `&'static str` for every filename referenced by a `Location`

We need to determine the impact to the ecosystem for each of these before deciding on mitigation strategy.

(1) was the main concern in the RFC but it was predicated on the assumption that `#[track_caller]` would affect inlining (an artifact of the implementation proposed). Since it doesn't affect inlining now, the only lost optimization in panicking branches would be for cases where LLVM previously used knowledge of a single constant panic location to make the failure path of a function smaller.

We've not yet had any reports of this and some ad-hoc measurements don't show any difference. This matches my mental model of LLVM's ability to optimize calling conventions which is "very expensive magic". I'm open to ideas for how to better assess this, including "wait and see".

**TLDR** Re (2) and (3),

`Location`s are 24 bytes when `size_of::<usize>() == size_of::<u64>()` and we don't currently measure how many we encode. We should put these in their own section to make it easier to measure.

Which filenames do we encode? In practice: probably all source files. We should put these in their own section too.

**Comment by anp** [comment]
@eddyb could you provide some pointers on how to put static allocations in their own sections? I assume it's a change to `librustc_codegen_ssa`?

**Comment by eddyb** [comment]
@anp Yeah, look at how `link_section` is handled and do the same.

**Comment by Mark-Simulacrum** [comment]
Have we looked at a diff of a stage1 (or stage2) rustc build? rustc has a lot of expect/unwrap etc code so I suspect if there's essentially no delta then I'm not too worried. I haven't noticed nightlies getting much bigger so I suspect we in practice aren't growing by too much.

**Comment by eddyb** [comment]
@Mark-Simulacrum this reminds me that we're not measuring size changes on perf.rust-lang.org.

**Comment by anp** [comment]
Filed: https://github.com/rust-lang/rustc-perf/issues/643

**Comment by anp** [comment]
OK, here's a [gist documenting an instrumentation-based approach](https://gist.github.com/anp/2137aec0bb2aa6a8cd6d70f5d6cea762). TLDR: rustc with `#[track_caller]` in the standard library uses ~0.68% of its binary size on the static `Location` allocations but the binary is still quite a bit smaller than a matching stage1 build. I haven't yet been able to observe a conclusive net increase in binary size due to the feature.

The fact that intuitively we know this *could* cause a net binary bloat makes me think that we want to polish up the custom sections patch in my gist. It's useful instrumentation for the future.

I'm not sure whether this is sufficient motivation to block stabilization on https://github.com/rust-lang/rust/issues/70580. Thoughts?

**Comment by tmandry** [comment]
The best way of calculating the impact of the feature itself would be doing two stage2 builds, one with track_caller turned on and one with it off. From your analysis now we know it's no more than 810 kB for rustc.

Percentage of the overall binary size is probably useful, but it's hard to say without seeing more binaries. For a large enough system I'd guess it'd be approximately the same overall, but for smaller binaries it could take up a higher percentage.

**Comment by benpye** [comment]
I think it would be interesting to allow putting the `CallerLocation` allocations in a different section. My thinking is that I would log the pointer to the `core::panic::Location` structure which would be in a no-load section of the binary - and so would be stripped from the flash image. A post processing tool could them take the log messages + the unstripped ELF and give you the original filename:line:column.
